### ISSUE TITLE
CA2016 bugs - Diagnostics triggered in edge cases

### ISSUE THREAD

**Issue by carlossanlop** [dotnet/roslyn-analyzers#3786]
The [dotnet/runtime PR](https://github.com/dotnet/runtime/pull/37607) to fix CA2016 in the repo source code found some edge cases that need to be fixed:

### Diagnostic triggered unexpectedly when there is an extra middle parameter

See https://github.com/dotnet/runtime/pull/37607/files#r445874631

The user was originally callin `TryAdd(T item)`, and the analyzer suggested to change it to use the one that takes a token. The expected behavior was that the user would not get a suggestion because the number of parameters differs by more than one between the two overloads.

```cs
        public bool TryAdd(T item)
        {...}
        public bool TryAdd(T item, TimeSpan timeout)
        {...}
        public bool TryAdd(T item, int millisecondsTimeout)
        {...}
        public bool TryAdd(T item, int millisecondsTimeout, CancellationToken cancellationToken)
        {...}
```

Similar thing happened in that same file with `TryTake`.

---

### Diagnostic triggered when an invocation inside a lambda that does not pass a ct, but the parent method does

See https://github.com/dotnet/runtime/pull/37607/files#r446107092

If the invocation that accepts an ct lives inside a lambda that does not pass a ct, do not offer a diagnostic.

```cs
using System;
using System.Threading;
public static class Extensions
{
    public static void Extension(this bool b, Action<int> action) { }
    public static void MyMethod(this int i, CancellationToken c = default) { }
}
class C
{
    public void M(CancellationToken ct)
    {
        bool b = false;
        b.Extension((j) =>
        {
            j.MyMethod();
        });
    }
}
```

---

### Static local method should not get a ct passed

```cs
using System;
using System.Threading;
class C
{
    public static void MyMethod(int i, CancellationToken c = default) {}
    public void M(CancellationToken ct)
    {
        LocalStaticMethod();
        static void LocalStaticMethod()
        {
            MyMethod(5);
        }
    }
}
```
